### PR TITLE
BUG: `np.lib.index_tricks.r_` mutates its own state

### LINKED ISSUE #8815: numpy.r_ seems to cache string beyond failing call
Consider the following code:

```python
from __future__ import print_function

import numpy as np

a = np.ones((3, 3, 3))
b = np.zeros((3, 3))

# Uncommenting this will make the below call to `np.r_` work.
# try:
# 	np.r_['2,3', a, b]
# except ValueError:
# 	print('failure')

print(np.r_[a, b.reshape((3, 3, 1))])
```

This fails with:

        Traceback (most recent call last):
      File "test.py", line 13, in <module>
        print(numpy.r_[a, b.reshape((3, 3, 1))])
      File "/usr/local/lib/python2.7/dist-packages/numpy/lib/index_tricks.py", line 338, in __getitem__
        res = _nx.concatenate(tuple(objs), axis=self.axis)
    ValueError: all the input array dimensions except for the concatenation axis must match exactly

All right as the dimensions don't match indeed. However uncommenting the above `try/except` block will make the second call to `numpy.r_` work, just as it was `numpy.r_['2,3', a, b.reshape((3, 3, 1))]` instead (stacking along the last axis). So it seems that `numpy.r_` still has that string from the previous failing call cached and is using it for the second call.

    failure
    [[[ 1.  1.  1.  0.]
      [ 1.  1.  1.  0.]
      [ 1.  1.  1.  0.]]

     [[ 1.  1.  1.  0.]
      [ 1.  1.  1.  0.]
      [ 1.  1.  1.  0.]]

     [[ 1.  1.  1.  0.]
      [ 1.  1.  1.  0.]
      [ 1.  1.  1.  0.]]]

Tested on:
* Python 2.7.12 + numpy==1.12.0
* Python 3.5.2 + numpy==1.12.0

### PR DESCRIPTION
**Pull request by eric-wieser**
Well, turns out that `np.r_` is buggy in all sort of exciting ways concerning that magic first argument.

Also, another case of _"`np.some_func` is broken and it turns out that `np.ma.some_func` has the same bugs but also the last few years of bug-fixes didn't get applied there"_

Fixes #8815

### CODE REVIEW

File: numpy/lib/index_tricks.py [hunk 1]
```diff
@@ -238,45 +238,41 @@ class AxisConcatenator(object):
...
-    def _retval(self, res):
-        if self.matrix:
-            oldndim = res.ndim
-            res = makemat(res)
-            if oldndim == 1 and self.col:
-                res = res.T
-        self.axis = self._axis
-        self.matrix = self._matrix
-        self.col = 0
-        return res
```
Comment (eric-wieser) [inline numpy/lib/index_tricks.py]: Problem was that this cleanup wouldn't run if there was an error. Better to not break things to the point that they need cleanup in the first place.

Also, this means that `np.r_` wasn't threadsafe. Hooray.

**Comment by eric-wieser** [comment]
I haven't added a test here, because "testing things have no mutable state when there's no reason they should" is a horrible thing to test for, especially if your tests can run in any order

**Comment by eric-wieser** [comment]
Note that right now, none of the commits pass without the last one. If this is a concern, I can reorder them to put the `ma` commit first

**Comment by eric-wieser** [comment]
Can we get this into 1.13? This is a pretty nasty bug, as it means that a working 3rd-party library can be broken by a careless piece of user code.

File: numpy/lib/index_tricks.py [hunk 2]
```diff
@@ -290,37 +285,35 @@ def __getitem__(self, key):
...
+                    axis = int(item)
                     continue
                 except (ValueError, TypeError):
                     raise ValueError("unknown special directive")
-            elif type(key[k]) in ScalarType:
-                newobj = array(key[k], ndmin=ndmin)
-                scalars.append(k)
+            elif type(item) in ScalarType:
+                newobj = array(item, ndmin=ndmin)
+                scalars.append(len(objs))
```
Comment (charris) [inline numpy/lib/index_tricks.py]: This is different. What is the explanation?

File: numpy/ma/extras.py [hunk 3]
```diff
@@ -1440,60 +1440,7 @@ class MAxisConcatenator(AxisConcatenator):
     mr_class
 
     """
-
-    def __init__(self, axis=0):
-        AxisConcatenator.__init__(self, axis, matrix=False)
-
-    def __getitem__(self, key):
-        if isinstance(key, str):
-            raise MAError("Unavailable for masked array.")
```
Comment (charris) [inline numpy/ma/extras.py]: Is this no longer the case?

Comment (eric-wieser) [inline numpy/ma/extras.py]: Correct - this now reuses the code path from `np.r_`. Do you want me to add a release note?

File: numpy/lib/index_tricks.py (hunk 2 above)
Comment (eric-wieser) [inline numpy/lib/index_tricks.py]: This fixes the test case added in 42ea3b54d8f852c9545f826548e198fafea127b9 - `k` was not the number of objects, but the number of arguments, which could include the string argument

Comment (eric-wieser) [inline numpy/lib/index_tricks.py]: On [this line](https://github.com/numpy/numpy/pull/8816/commits/42ea3b54d8f852c9545f826548e198fafea127b9#diff-1bd953557a98073031ce66d05dbde3c8R337), we index `objs` with a value from `scalars`. Previously, there was an off-by-one error that at best resulting in incorrect casting, and at worse an AttributeError or IndexError

File: numpy/ma/extras.py [hunk 4]
```diff
@@ -1440,60 +1440,7 @@ class MAxisConcatenator(AxisConcatenator):
...
-                if final_dtypedescr is None:
-                    final_dtypedescr = newobj.dtype
-                elif newobj.dtype > final_dtypedescr:
-                    final_dtypedescr = newobj.dtype
-        if final_dtypedescr is not None:
-            for k in scalars:
-                objs[k] = objs[k].astype(final_dtypedescr)
-        res = concatenate(tuple(objs), axis=self.axis)
-        return self._retval(res)
+    concatenate = staticmethod(concatenate)
```
Comment (charris) [inline numpy/ma/extras.py]: Doing this with the decorator might be preferable. I assume that here `concatenate` is a masked version and the only thing that needed changing?

Comment (eric-wieser) [inline numpy/ma/extras.py]: Correct - doing a diff between the two showed `np.ma.concatenate` vs `np.concatenate` to be the only difference that wasn't the result of a forgotten feature/bugfix.

You want me to change this to the much more verbose:

```
@staticmethod
def concatenate(arrs, axis=-1):
    return concatenate(arrs, axis)
```

File: numpy/ma/extras.py (hunk 3 above)
Comment (eric-wieser) [inline numpy/ma/extras.py]: Actually, you're right - this case makes little sense for masked arrays

Comment (charris) [inline numpy/ma/extras.py]: Release note would be good. I assume something else uses this class and has a docstring that should be updated also. The base class itself should probably get a docstring at some point, maybe the baseclass should even be made an ABC, but that isn't needed for 1.13.

File: numpy/ma/extras.py (hunk 4 above)
Comment (charris) [inline numpy/ma/extras.py]: I have no strong feelings one way or the other, let's let it be. I was mostly reacting to the decorator being the more modern method.

File: numpy/ma/extras.py (hunk 3 above)
Comment (eric-wieser) [inline numpy/ma/extras.py]: Actually,this path makes no sense for masked arrays, since it returns a plain np.matrix - I'll add back the check

**Comment by charris** [comment]
> none of the commits pass without the last one.

I'm tempted ot just do a squash merge if that would be OK.

File: numpy/ma/extras.py (hunk 3 above)
Comment (eric-wieser) [inline numpy/ma/extras.py]: Ok, this line is back

**Comment by eric-wieser** [comment]
> I'm tempted ot just do a squash merge if that would be OK.

Rebase incoming

File: numpy/ma/extras.py [hunk 5]
```diff
@@ -1461,60 +1461,19 @@ class MAxisConcatenator(AxisConcatenator):
...
-    def __init__(self, axis=0):
-        AxisConcatenator.__init__(self, axis, matrix=False)
+    @staticmethod
+    def makemat(arr):
+        return array(arr.data.view(np.matrix), mask=arr.mask)
 
     def __getitem__(self, key):
+        # matrix builder syntax, like 'a, b; c, d'
         if isinstance(key, str):
             raise MAError("Unavailable for masked array.")
```
Comment (eric-wieser) [inline numpy/ma/extras.py]: Delegating wouldn't work here anyway, as it would end up looking for locals in this stack frame, not the user's stack frame.

Comment (eric-wieser) [inline numpy/ma/extras.py]: In general, I think it would be nice to deprecate `bmat('a, b; c, d')`...

**Comment by eric-wieser** [comment]
Alright, the rebase passes, and the comments you had are addressed.

I don't think there's anything releasenote worthy here, as everything is a bug fix

**Comment by charris** [comment]
Thanks Eric.

**Pull request merged**